Shipping Labels: Print customs forms

itsmeichigo opened the pull request:
Closes #4597

# Description
This PR creates new view for printing customs form, and redirect to this view from order details and shipping label printing view if the label has a separate customs invoice.

# Changes
- Add new view `PrintCustomsFormsView`. This view was set up to accept more than one invoice to support multi-package later.
- Updated Order Details screen to show option to print customs form from a shipping label section if the label has `commercialInvoiceURL`.
- Updated `PrintShippingLabelCoordinator` to show the print view for customs form in the print flow if the current label has an invoice.

# Demo
| From Order Details | From Print flow |
| ----- | ----- |
| ![print-customs-1](https://user-images.githubusercontent.com/5533851/131101814-234c4fe0-7e02-4fb6-bf4a-f7356d460680.gif) | ![print-customs-2](https://user-images.githubusercontent.com/5533851/131101855-fd81f19b-6ff5-4423-a848-02b7f226862b.gif) |

# Testing
1. Make sure that your test store has installed WCShip plugin and configured packages including DHL packages in WPAdmin > WooCommerce > Settings > Shipping > WooCommerce Shipping.
2. Create an international order (origin and destination countries are different).
3. Navigate to Orders tab on iOS app, select the new order and select Create Shipping Label.
4. Configure Ship From, Ship To, Package Details and Customs. Please make sure to enter phone number for both origin and destination addresses.
5. In Carriers and Rates, select DHL as the carrier.
6. Proceed to purchase the label.
7. After the purchase succeeds, select a paper size and select Print Shipping Label.
8. Close the print view, notice that customs forms printing view is displayed. Tapping on Print Customs Form should navigate to a print view, tapping on Save for Later should dismiss the view.
9. Head back to Order Details screen, select the overflow menu in the Shipping Label section. Notice that there's an option for Print Customs Form in the presented action sheet.
10. Select the option, notice that the same customs form printing view is displayed.

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Add PrintCustomsFormsView

itsmeichigo committed:
Update PrintCustomsFormsView with buttons

itsmeichigo committed:
Add constants for print customs forms view

itsmeichigo committed:
Add localization for print customs form view

itsmeichigo committed:
Make print customs forms view scrollable

itsmeichigo committed:
Show button on order detail to navigate to print customs forms view

itsmeichigo committed:
Handle dismiss print customs forms view when tapping Save for later

itsmeichigo committed:
Handle printing for print customs forms view

itsmeichigo committed:
Remove dismissHandler for print customs forms view

itsmeichigo committed:
Update position for print rows of print customs forms view

itsmeichigo committed:
Remove printHandler for print customs forms view

itsmeichigo committed:
Show customs form printing after printing label

itsmeichigo committed:
Update showPrintingView method to accept string

itsmeichigo committed:
Show InProgressView when downloading customs document for printing

pmusolino reviewed (changes requested):
I tested it, but I found some issues.

1) I found a small bug about the image of the FedEx courier. I created an issue here https://github.com/woocommerce/woocommerce-ios/issues/4888 since it's not related to this PR.
2) Tapping on Print Customs Form navigate to the print view, but during the loading, the screen is not attached to the borders of the view, but to the safe area.
<img src="https://user-images.githubusercontent.com/495617/131113824-84dbf58b-f2f2-4771-ad81-808a34a88786.png" width=300 />
3) I wasn't able to test step n°9. `Head back to Order Details screen, select the overflow menu in the Shipping Label section. Notice that there's an option for Print Customs Form in the presented action sheet.`. Basically, I'm not able to see the shipping label purchased. 🤔

itsmeichigo committed:
Make InProgressView ignore safe area on PrintCustomsFormsView

itsmeichigo committed:
Merge branch 'develop' into task/4597-customs-form-printing-view

itsmeichigo commented:
Thanks @pmusolino for spotting the issues! I have fixed the safe area issue with the progress view in 12d05eb0a5b23d48434ab60cacea1c2c237996ba.

As for step 9, I'm fairly positive this was due to the failure in parsing purchased labels, which I fixed in #4886. I have merged the PR and merged develop back into this branch so can you try again with the order that you could not find the purchased labels?

itsmeichigo committed:
Hide Save For Later button from customs printing view by default

itsmeichigo committed:
Update source UI for PrintShippingLabelCoordinator navigation controller

itsmeichigo committed:
Update unit test for PrintShippingLabelCoordinator

itsmeichigo committed:
Remove Shipping Label Print controller from navigation stack before pushing invoice customs printing

itsmeichigo committed:
Put check for printType back

itsmeichigo committed:
Only show customs printing after completing printing shipping label

itsmeichigo commented:
**Update** - I've pushed quite a few changes after testing with Android app:
- Save for Later button is only present in print flow and not reprint flow.
- For print flow: navigation stack is updated to remove Shipping Label print view before pushing Customs Invoice print view. This is to make sure that when tapping back or Save for Later button, user is redirected to Order Details screen immediately. Internal reference: p1622636091163700/1622543318.153700-slack-CGPNUU63E.
- Only show Customs Invoice printing in print flow after completing printing label. Android is showing the screen regardless since they cannot detect if printing is completed; but since we can get this from the completion block of `UIPrintInteractionController`, we should use it. Internal reference: p1622636072163500/1622543318.153700-slack-CGPNUU63E.

pmusolino commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Print Customs Form/PrintCustomsFormsView.swift:
```diff
@@ -0,0 +1,165 @@
+import SwiftUI
+
+struct PrintCustomsFormsView: View {
+    let invoiceURLs: [String]
```
nit 
```suggestion
    private let invoiceURLs: [String]
```

pmusolino commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Print Customs Form/PrintCustomsFormsView.swift:
```diff
@@ -0,0 +1,165 @@
+
+struct PrintCustomsFormsView: View {
+    let invoiceURLs: [String]
+    let showsSaveForLater: Bool
```
nit 
```suggestion
    private let showsSaveForLater: Bool
```

pmusolino commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Print Customs Form/PrintCustomsFormsView.swift:
```diff
@@ -0,0 +1,165 @@
+        }
+    }
+
+    private var saveForLaterButton: some View {
+        Button(action: {
+            presentationMode.wrappedValue.dismiss()
+        }, label: {
+            Text(Localization.saveForLaterButton)
+        })
+        .buttonStyle(SecondaryButtonStyle())
+        .renderedIf(showsSaveForLater)
+    }
```
I like this approach to separate small views into private sub-components. The view becomes more readable, and with less nesting. 💯

itsmeichigo replied:
Actually I extract this view into a separate component since it's used in 2 places and I don't want to repeat myself 😅. But yeah, I'm trying to keep the balance between keeping `body` flat while avoiding scattering too many small components since it can be hard to follow.

pmusolino reviewed (approved):
Thank you @itsmeichigo for fixing all the things, now it works perfectly :shipit:

itsmeichigo committed:
Keep properties private for PrintCustomsFormsView
